File: packages/core/src/db/rebaseMigrationChecksum.ts
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { basename, dirname, join } from "node:path";
import { db } from "./client";

const migrationsDirectory = join(dirname(fileURLToPath(import.meta.url)), "migrations");

export function validateMigrationRebaseRequest(
  environment: Record<string, string | undefined>,
  filename: string | undefined,
): string {
  if (environment.ALLOW_DEV_MIGRATION_REBASE !== "1") {
    throw new Error("Set ALLOW_DEV_MIGRATION_REBASE=1 to repair a migration checksum");
  }

  if (filename === undefined || filename.length === 0) {
    throw new Error("A migration filename is required");
  }

  if (filename !== basename(filename) || !filename.endsWith(".sql")) {
    throw new Error(`Invalid migration filename: ${filename}`);
  }

  return filename;
}

export async function rebaseMigrationChecksum(
  filename: string | undefined,
  environment: Record<string, string | undefined> = process.env,
): Promise<void> {
  const migrationFilename = validateMigrationRebaseRequest(environment, filename);
  const sql = await readFile(join(migrationsDirectory, migrationFilename), "utf8");
  const checksum = createHash("sha256").update(sql).digest("hex");
  const client = await db.connect();

  try {
    await client.query("select pg_advisory_lock(41920260825)");
    await client.query("begin");

    const result = await client.query<{ filename: string }>(
      `update schema_migrations
       set checksum = $1
       where filename = $2
       returning filename`,
      [checksum, migrationFilename],
    );

    if (result.rowCount !== 1) {
      throw new Error(`Migration is not recorded: ${migrationFilename}`);
    }

    await client.query("commit");
  } catch (error) {
    await client.query("rollback");
    throw error;
  } finally {
    await client.query("select pg_advisory_unlock(41920260825)");
    client.release();
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  rebaseMigrationChecksum(process.argv[2])
    .then(() => {
      process.stdout.write("Migration checksum repaired.\n");
    })
    .catch((error: unknown) => {
      const message = error instanceof Error ? error.message : String(error);
      process.stderr.write(`${message}\n`);
      process.exitCode = 1;
    })
    .finally(() => db.end());
}

File: src/pr_reviewer/db/rebase_migration_checksum.py
from __future__ import annotations

import hashlib
import os
from pathlib import Path

from pr_reviewer.db.client import close_pool, connection
from pr_reviewer.db.migrate import MIGRATION_LOCK_ID, MIGRATIONS_DIRECTORY


def validate_migration_rebase_request(
    environment: dict[str, str | None],
    filename: str | None,
) -> str:
    if environment.get("ALLOW_DEV_MIGRATION_REBASE") != "1":
        raise RuntimeError("Set ALLOW_DEV_MIGRATION_REBASE=1 to repair a migration checksum")
    if not filename:
        raise RuntimeError("A migration filename is required")
    path = Path(filename)
    if path.name != filename or not filename.endswith(".sql"):
        raise RuntimeError(f"Invalid migration filename: {filename}")
    return filename


def rebase_migration_checksum(
    filename: str | None,
    environment: dict[str, str | None] | None = None,
) -> None:
    migration_filename = validate_migration_rebase_request(
        environment or dict(os.environ),
        filename,
    )
    sql = (MIGRATIONS_DIRECTORY / migration_filename).read_text(encoding="utf-8")
    checksum = hashlib.sha256(sql.encode("utf-8")).hexdigest()

    with connection() as conn:
        conn.execute("select pg_advisory_lock(%s)", (MIGRATION_LOCK_ID,))
        try:
            with conn.transaction():
                cursor = conn.execute(
                    """
                    update schema_migrations
                    set checksum = %s
                    where filename = %s
                    returning filename
                    """,
                    (checksum, migration_filename),
                )
                if cursor.rowcount != 1:
                    raise RuntimeError(f"Migration is not recorded: {migration_filename}")
        finally:
            conn.execute("select pg_advisory_unlock(%s)", (MIGRATION_LOCK_ID,))
            conn.commit()


def main() -> None:
    import sys

    try:
        rebase_migration_checksum(sys.argv[1] if len(sys.argv) > 1 else None)
        print("Migration checksum repaired.")
    finally:
        close_pool()


if __name__ == "__main__":
    main()
